libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,197 @@
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto output = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+	{
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	}
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+	{
+		output += collect(declaration, _sourceUnit, declaration->name());
+	}
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_result.emplace_back(Reference{symbolAlias.location, DocumentHighlightKind::Read});
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_result.emplace_back(Reference{symbolAlias.location, DocumentHighlightKind::Text});
+			break;
+		}
+	}
+}
+
+bool ReferenceCollector::tryAddReference(Declaration const* _declaration, SourceLocation const& _location)
+{
+	if (&m_declaration != _declaration)
+		return false;
+
+	m_result.emplace_back(Reference{_location, m_kind});
+	return true;
+}
+
+void ReferenceCollector::endVisit(Identifier const& _identifier)
+{
+	if (auto const* declaration = _identifier.annotation().referencedDeclaration)
+		tryAddReference(declaration, _identifier.location());
+
+	for (auto const* declaration: _identifier.annotation().candidateDeclarations + _identifier.annotation().overloadedDeclarations)
+		tryAddReference(declaration, _identifier.location());
+}
+
+
+void ReferenceCollector::endVisit(IdentifierPath const& _identifierPath)
+{
+	tryAddReference(_identifierPath.annotation().referencedDeclaration, _identifierPath.location());
+}
+
+void ReferenceCollector::endVisit(MemberAccess const& _memberAccess)
+{
+	if (_memberAccess.annotation().referencedDeclaration == &m_declaration)
+		m_result.emplace_back(Reference{_memberAccess.location(), m_kind});
```

Comment:
I guess the same goes for all `emplace_back` usages here; `emplace_back(Reference{...})` will create a temporary, and then `move` it, so this would I guess be equivalent to simply calling `push_back(Reference{...})` as it there's really no difference between `push_back(&&value_type)` vs `emplace_back(&&value_type)`.

Using `emplace_back(Args&& args...)` will construct in place, which is the preferred way of emplacing.

Reply:
Without having thought about it too much in the past, that is specifically why I prefer `emplace_back` as much as possible. But I remember that I had some issues with inplace arguments, and that's why I seem to have the common ground of doing `emplace_back(Type{...})`. I am sure the release mode optimizer will make it equivalent to `emplace_back(...)`. I'll change it regardless.

Reply:
I believe the optimizer will make it equivalent to `emplace_back(Args&& args...)` if you pass in `rvalue`s to the constructed object only, e.g. `emplace_back(Reference{rvalue, rvalue})`, but will otherwise make a temporary copy and then move it.